### ISSUE TITLE
Support optimizer.step() e2e on device - running optimization step for updating model parameters

### ISSUE THREAD

**Issue by nvukobratTT** [tenstorrent/tt-forge-fe#178]
## Summary
Optimizer is required for weights update once gradients are calculated. Effort of this issue should be to map `optimizer.step()` as a runtime API to run optimization graph based on runtime metadata got after MLIR compile and flatbuffer generation.

## Details
Most of the optimizer related details can be found here tenstorrent/tt-forge-fe#176. In sum, optimizer should be separated flatbuffer as well (MLIR only sees the graph; not aware of training specifics). 

Key goal of this issue is to explore how we can map `optimizer.step()` to run optimizer flatbuffers. The flow can work like this:
1. Initialize model (fwd grph)
2. Mark it as training
3. Compile the model. It's important to compile model before optimizer (details below)
4. Compile the optimizer
5. Run training loop with optimizer

Few key points to note here are:
- For 3. point, we should enforce that model is compiled first. In order for optimizer to be able to map gradient accumulation inputs, and parameter outputs, runtime has to compile model first in order to get that informatin
- For 4. point, but similar to above mentioned point, optimizer needs to be aware of valid inputs and outputs. In sum, backward graph will contain information about where are gradients stored/accumulated. Before and after model compilation, runtime needs to keep parameter references and match those correctly: which grad accumulation input maps to which model parameter.

## Passing criteria
- Optimizer is separate flatbuffer
- Optimizer has a valid grad accumulation inputs mapped to model parameters
- Gradients are successfully updated
- Updated gradients (after X steps) are matching CPU (Golden)
